### PR TITLE
Add Rollkit CLI

### PR DESCRIPTION
**Pull request by Ferret-san**
Closes: #745

## Key Update

Most notably this PR adds a `rollkit` CLI tool. Here are some of the changes that come along with it. 

### CLI Docs

A `docs-gen` command was added to the CLI to auto generate the documentation for the CLI tool. This documentation likes in `./cmd/rollkit/docs`. 

These Markdown files are then ignore by the markdown linter.

### Makefile changes

Currently rollkit is used via the cosmos-sdk. The outdated reference to `go build` in the README was removed as there are now current requests for using a rollkit binary. 

This allows for the default `make install` to be for the rollkit CLI. 

### Commands

The initial commands of the CLI are:
- `docs-gen`
- `start`
- `version`

`docs-gen` has been previously described. 

`rollkit start` is the primary command. It handles initialization as a pre-run step and launches the node for the user. 

`version` is used to see the current version that the rollkit CLI is using. This is dynamically set from the git history. 

----------

## Summary by CodeRabbit

- **New Features**
	- Introduced a modular framework for rollups with a comprehensive CLI for configuration, logging, and error handling.
	- Added autocompletion script generation commands for bash, fish, PowerShell, and zsh shells.
	- Implemented a `rollkit start` command to run the rollkit node with extensive configuration options.
	- Provided a `rollkit version` command to display version information for Rollkit.
	- Introduced documentation generation functionality for the rollkit CLI.
- **Documentation**
	- Added detailed guides for setting up autocompletion in various shells.
	- Introduced documentation on running the rollkit node and displaying version information.
- **Chores**
	- Implemented a new lint job for CLI documentation to ensure quality and consistency.

### CODE REVIEW

**Comment by gupadhyaya** [comment]
@Ferret-san could you resolve the conflicts for reviewing?

File: cmd/README
Comment (tzdybal) [inline cmd/README:1]: More description please

**Review by tzdybal — COMMENTED** [review]
I'm leaving some comments.
I agree that this needs to be rebased before more reviews, especially after resolution of appClient/proxyApp PR.

File: config/priv_validator_key.json
Comment (Ferret-san) [reply config/priv_validator_key.json:1]: oops :rofl:

File: cmd/rollkit/commands/run_node.go
Comment (Ferret-san) [reply cmd/rollkit/commands/run_node.go:90]: you mean we should make consts for the default values in these flags? (like default block time, etc)

Comment (nashqueue) [reply cmd/rollkit/commands/run_node.go:90]: I think he means we should make constants in the confi.go and then export them and let them be used here. This would ensure consistency.

File: cmd/README
Comment (smuu) [inline cmd/README:1]: It would be nice to have an explanation and examples of how to use the CLI.

File: types/pb/tendermint/abci/types.pb.go
Comment (nashqueue) [inline types/pb/tendermint/abci/types.pb.go:1]: Can you delete this file again? It should not get pushed as it is in gitignore now

File: go.mod
Comment (Ferret-san) [reply go.mod:1]: I used cometbft for this since the cometbft would come after the cli, and by having it be on cometbft already we save ourselves the work of updating it.

Comment (Ferret-san) [reply go.mod:1]: but as you mention, we are only using it once, so it's ok to switch it to tendermint for the brief time period before cometbft gets merged

File: types/pb/dalc/dalc.pb.go
Comment (gupadhyaya) [inline types/pb/dalc/dalc.pb.go:1]: prefer not modifying the generated files. can you remove them?

**Comment by gupadhyaya** [comment]
overall looks good. we can merge this after you clean-up the commented parts and extend README.md with some more examples.

**Review by tzdybal — CHANGES_REQUESTED** [review]
This is really cool.

I left some comments & change requests.

File: cmd/rollkit/main.go
Comment (smuu) [inline cmd/rollkit/main.go:1]: Does it make sense to build a container image for that?

Comment (nashqueue) [reply cmd/rollkit/main.go:1]: A rollkit-only container without the app/cosmos SDK ? and then a separate container with only the app ?

Comment (smuu) [reply cmd/rollkit/main.go:1]: Mainly for the setup when having an execution environment that is ABCI compatible.
Having rollkit and the cosmos-sdk app separated is a separate discussion, in my opinion.

File: cmd/rollkit/commands/run_node.go
Comment (Ferret-san) [reply cmd/rollkit/commands/run_node.go:66]: lol, this is straight from tendermint cli :rofl:

File: cmd/rollkit/commands/run_node.go
Comment (Ferret-san) [reply cmd/rollkit/commands/run_node.go:32]: perhaps, iirc I either got this from something else on rollkit, or these where values I used at some point for testing

File: cmd/rollkit/commands/run_node.go
Comment (Ferret-san) [reply cmd/rollkit/commands/run_node.go:138]: Ah mb, those prints are just me being paranoid and doing some debugging, will get rid of them

File: cmd/rollkit/commands/init.go
Comment (Ferret-san) [reply cmd/rollkit/commands/init.go:62]: no clue tbh, but I'm down to change it if it makes sense (this is something I got from the tendermint cli)

File: cmd/rollkit/commands/init.go
Comment (Ferret-san) [reply cmd/rollkit/commands/init.go:44]: I beliebe tendermint does this for more detailed logging, but it might make sense to change it

File: cmd/rollkit/commands/root.go
Comment (Ferret-san) [reply cmd/rollkit/commands/root.go:88]: need to revisit this

File: cmd/rollkit/commands/init.go
Comment (Ferret-san) [reply cmd/rollkit/commands/init.go:25]: another case of "I just got this from tendermint cli", can def get rid of it

File: cmd/rollkit/commands/version.go
Comment (Ferret-san) [reply cmd/rollkit/commands/version.go:14]: yeah I think keeping it the same as the rollkit version would be best

File: cmd/rollkit/commands/run_node.go
Comment (Ferret-san) [reply cmd/rollkit/commands/run_node.go:32]: 30s is the default block time provided by `DefaultNodeConfig`

**Comment by nashqueue** [comment]
@coderabbitai resume

**Comment by MSevey** [comment]
@jcstein would love your thoughts on features / functionality this should have out the gate as we go about updating this.

File: cmd/README.md
Comment (gupadhyaya) [inline cmd/README.md:7]: i think go version requirement is higher `>= 1.21`

File: config/config.go
Comment (gupadhyaya) [inline config/config.go:18]: no more da_layer flag.

File: config/config.go
Comment (gupadhyaya) [inline config/config.go:20]: you can remove this too.

File: cmd/rollkit/rollkit_start.md
Comment (gupadhyaya) [inline cmd/rollkit/rollkit_start.md:21]: you can update the flags here too.

File: cmd/rollkit/commands/run_node.go
Comment (gupadhyaya) [inline cmd/rollkit/commands/run_node.go:99]: use `defaultConfig` instead of `cmCfg.DefaultConfig()`

**Review by gupadhyaya — APPROVED** [review]

**Pull request merged**
